### ISSUE TITLE
Create an Action for CI/CD of Model Documents and Files for Sci Community Participation

### ISSUE THREAD

**Issue by jamesmrollins** [lzim/teampsd#1432]
### Check off requirements to verify progress through feature_tracker.
- [x] 1. **DEFINE in work_breakdown**
  - [x] a. 

**HYPOTHESIS**
-  A CI/CD action that will provide automated services to produce SDM-DOCs and XMILE compatible model files to the MTL Demo page will provide scientific transparency and usefulness to users in the scientific community, who want to build on Team PSD's work. 

**ASSUMPTIONS**
-  The system dynamics approach depends on model transparency.  Transparency, or the quality of being easily observed and comprehended is an important characteristic of useful models because it helps users to identify the, "Assumptions, relationships and data used in the model." ( [Martinez-Moyano, 2012, accessed 7/20,](https://onlinelibrary.wiley.com/doi/full/10.1002/sdr.1471) )
- The System Dynamics Model Documentation and Assessment Tool (SDM-Doc) can provide documentation of models built using Vensim, is open source and can be adapted into the GitHub CI/CD process.
-  Interoperable model formats, such as XMILES, are important to support the use of a system dynamics standard, the development of common-use libraries, integration with big-data, and promote the integration of, "system dynamics models and simulations into mainstream analytics software." (accessed 7/20, [https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=xmile#press](https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=xmile#press) )

What this means:  If a user updates the model directory in the Sim UI repository, an action would automatically update the SDM-DOC on the MTL DEMO website. It would also provide access to XMILE model files for use in other model execution apps or platforms.  This task is related to #1220 Automatic Model Update Process.

  - [ ] b. **date & who** -  **Verified** requirement definitions an **existing team meeting.**
  
  Task Cards:
  #
---
- [ ] 2. **DESIGN** in **operations_to_do** and **research_to_do**
  - [ ] a. **what & why** - **_MVP_ of design** of requirements in a **algorithm map, wireframes or design** on the **master_feature_card.**
  - [ ] b. **date & who** -  **Verified** design **asynchronously** or an **existing team meeting.**
  
  Task Cards:
  #
---
- [ ] 3. **DEVELOP** in **under_development** - "frozen"

  Task Cards:
  #
---
- [ ] 4. **TEST in functional_test**
  - [ ] a. **what & why** - **TEST requirements** using **testing_checklist** on the **master_feature_card.**
  - [ ] b. **date & who** -  **Verified** via TEST **asynchronously** or an **existing team meeting.**
  
  Task Cards:
  #
---
- [ ] 5. **PROD in done**
  - [ ] a. **what & why** - **_DOCUMENT_ feature** using **document_tracker** for the **5 D's.** 
  - [ ] b. **date & who** -  **DELEGATE**
    - [ ] **describe_learners** [Jenn & Rita]
    - [ ] **detail_facilitators** [Jane & Debbie]
    - [ ] **document_team** [Anthony]
    - [ ] **disseminate_scientists_va** [Anthony]
    - [ ] **depend_products** [hq & ees]
  
  Task Cards:
  #
---
- [ ] 6. **closed** - fully documented, trained and ready for the next release.

**Comment by jamesmrollins** [comment]
Hi @lzim @dlounsbu @dkngenda @anthonycpichardo .  Below is the configuration table for the SDM-Doc Tool Version 5.0.  The HTML files located here,  are the outputs for the Care Coordination Vensim model.  I need to know what configuration you need and which outputs are needed.  Do you need them all?  Many seem redundant.  

![image](https://user-images.githubusercontent.com/30132017/97338945-353b0a00-183f-11eb-92dc-f4d307239580.png)

**Comment by lzim** [comment]
Thanks @jamesmrollins  I'll look at the sample at the link to see.

**Comment by lzim** [comment]
@jamesmrollins 

Do you know what the differences between these are expected to be? 
Or, is this just a look at each of them to see?

![Screen Shot 2020-10-27 at 1 32 07 PM](https://user-images.githubusercontent.com/7454688/97358719-e4d0a600-1858-11eb-8a65-7080eda8fdae.png)

**Comment by lzim** [comment]
@jamesmrollins  

Thanks for the quick turn around, as far as I can tell, these SDM-DOC reports contain the same information, but what differs about them is simply how the information is ordered/displayed. If that makes sense to you @jamesmrollins, then I think any of these output reports should be fine for transparency and reporting. A user of the documentation can use the hyperlinks in the .html file to navigate and adjust the views anyway, which I've always appreciated about SDM-DOC.

So, we want to produce the same type of SDM-DOC report for each of our 5 models and we'll include them at our main **mtl.how** repo here: https://github.com/lzim/mtl/tree/master/blue/mtl2.0_models

If possible, my preference would be that the file name would correspond to the model, but include 'sdm_doc' at the end, so that the files will be sorted in order, like this:

"cc_v36_2019_09_06.mdl" 
"cc_v36_2019_09_06_sdm_doc.html"

@dlounsbu @dkngenda @anthonycpichardo @staceypark

**Comment by jamesmrollins** [comment]
Hi @lzim 
WILCO on your instructions above.  One additional question - do you want to replace the .mdl Vensim model files with .xmile format model files in the repository?  XMILE is supposed to be standard and executable, and it human-readable like the .mdl file.

**Comment by lzim** [comment]
Thanks @jamesmrollins 

1. It see it sounds like I had a variety of view functionality because all the versions of the files were in the same working directory. May need to isolate to determine whether/how to narrow down at all from adding all .html.

2. Great question. As you know, I am VERY interested in making the .xmile available. It does make sense to me that we could have for each model version and date, these three in mtl.how/blue/mtl_2.0:
- .html (TBD whether we can reduce the # of .html).
- .mdl
- .xmile

3. Let's hold on this question. Short run, it's more important that we get the 1 & 2 on mtl.how and available for reporting in manuscripts/report. Mid run, it's more important that we get the CI/CD functionality for making changes throughout (e.g., changing a module name from "aggregate" to "team").

**Comment by lzim** [comment]
@jamesmrollins 

I isolated one SDM-DOC .html report (see screenshot of "**cc_i_v36_2019_09_06-A.html**) by itself into one working directory and although it did not dynamically re-organize upon clicking the first 5 blue hyperlinks (type, module, group, module/group/name, view), the "i" file did re-organize for the last two purple  "Level Structure" and "view summary).

As far as I can tell:
- Each report contains all the same information.
- Given this, and the fact that
   a. I don't know of a preference for specific ways of sorting the information, and
   b. I don't know which exact combinations are available for each file, with each of these endings:
      - A
      - G
      - M
      - MGA
      - T
      - V
   c. **I propose we export an -A file for each of our 5 models, and post with the .mdl and .xmile files**  

![Screen Shot 2020-10-29 at 2 49 51 PM](https://user-images.githubusercontent.com/7454688/97641996-682cfb80-1a01-11eb-969e-c65d8b33824a.png)
and

**Comment by jamesmrollins** [comment]
@lzim a pull request has been entered into lzim/mtl for the files noted above as instructed, for review by @lijenn and @staceypark.

**Comment by jamesmrollins** [comment]
@lzim for this epic, we will accomplish the CI/CD deployment of the SD-DOC and XMILE files  to lzim/mtl.  However, I will create a new issue to manage the deployment of the SD-DOC to the MTL DEMO website for the next epic, so we can close this one when tested and complete.  I will add the new issue cross reference here when complete.

**Comment by jamesmrollins** [comment]
Hi @lzim

### BLUF
Vensim cannot reliably export .xmile versions of our models.  Please accept recommendations below.

### FACTS
1.  There are commands in Vensim that are not supported in .xmile.  When the model attempts to export in this case, the Vensim application "crashes."  It really shouldn't, and Ventana is trying to get to the bottom of it.
2.  The export problem is intermittent.  It doesn't always fail; nevertheless, it is not reliable and capable.
3.  I have reported the issue to Ventana, who is working to resolve in the next release (see exhibit below).

### RECOMMENDATION
1.  Continue to update models using .mdl format and provide SDM-DOC as designed.
2.  When Vensim releases update, we can add the auto-update XMILE function.
3.  Since the models will likely not change for several months, the current XMILE saves should work fine.
4.  I will separate the XMILE auto-update function as a separate issue and put in the EPIC BACKLOG pipeline so we don't loose visibility. 


**Exhibit**
------------------------------------------
 From: Tony Kennedy - Ventana <vensim@vensim.com>

I understand.

For transparency, XMILE has benefits and drawbacks. If working for the VA, it might make more sense to stick with MDL format models. Many governments ask for the VPAT document for Vensim when purchasing (it details what Vensim does so it can be used by the physically impaired to create models in plain text). XMILE is not really plain in the way that a MDL file is.

 >> With the intermittent nature of the crash, I am not sure I understand the root cause of the crash, which I need to fix in order to develop a reliable  >> automatic routine. If I understand your comment below, the root cause is an unsupported command in the Vensim model code, which causes the  >> conversion to crash?  If unsupported commands are the root cause, is there a list of unsupported commands and substitutions that I can refer to?
It shouldn't crash, that's our fault and we need to fix it. The main problem is that XMILE is not really well used at all (which means very limited testing). In this case, it's crashing because there is an equation combination we haven't seen before (STEP embedded within a DELAY function, the parser gets confused). We implemented the standard features of XMILE, but stayed away from adopting it as a native file format due to the VPAT requirements. We would much prefer to keep a plain text format model that is easily interpreted by text to speech tools. Hope this makes sense, we are being asked a lot more regularly now for the VPAT document.

For now, I cannot really get you a fix or a workaround for your model. 
The next release of Vensim will have the fixes, are you able to use MDL format models until then? They are much easier to read than XMILE (apart from the sketch). The other alternative would be to use the document tool to generate a model report.

All the best,

Tony.

**Comment by jamesmrollins** [comment]
Hi @lzim @staceypark @lijenn @TomRust @dkngenda @dlounsbu 
FYI: @hirenp-waferwire 

### PROBLEM
Vensim ver. 8.0 cannot be run from a GitHub repository.  This is a hard constraint.  Version 8.0 is necessary for an XMILE conversion.

### BLUF
We can update Model Documents and automatically generate multiple file versions and deploy to given locations in Epicenter using a client/web hybrid process.   

### FACTS
1.  Working with Tony Kennedy from Ventana, he provided an upgraded copy of Vensim version 8.0.  This Vensim version will successfully convert a .mdl to .xmile and binary (.vmf).
2. Modelers must upgrade to v 8.0 if they haven't already done so.
3. Modelers must download GitHub desktop or equivalent and clone the lzim/mtl repo.  This will give them the ability to access model files from their desktops.
4. DEV can provide necessary batch processing files that will make necessary conversions to .xmile and binary formats from the client desktop.

### SOLUTION
1.  After the modeler clones the repository using GitHub Desktop, they will be able to access the model files directly from their client desktop using their Vensim license.
2.  When the modeler makes changes and commits to the DEV branch in the GitHub Desktop, the batch process will make the necessary conversions (.vmf, .xmile and SD-DOC). 
3.  When the modeler pushes the changes back to origin, the GitHub origin (web) will execute UploadModel action (see #1220), which will deposit files in the necessary Epicenter projects.

**Comment by jamesmrollins** [comment]
This action is complete and dependent upon #1637 for deployment to PROD.
